examples/common/load_dds.adb

```diff
@@ -0,0 +1,277 @@
+
+    exception
+        when others =>
+            Put_Line ("An exception occurred in Load_DSS_Header.");
+            raise;
+    end Load_DSS_Header;
+
+    --  ------------------------------------------------------------------------
+
+    procedure Load_Mipmaps (Header                        : tDDS_Header; DDS_Data : tDDS_Data;
+                            Block_Size                    : GL.Types.UInt;
+                            Initial_Width, Initial_Height : GL.Types.UInt;
+                            Format                        : Gl.Pixels.Internal_Format) is
+        use GL.Objects.Textures;
+        use GL.Types;
+        Width        : UInt := Initial_Width;
+        Height       : UInt := Initial_Height;
+        Mip_Size     : UInt := ((Width + 3) / 4) * ((Height + 3) / 4) * Block_Size;
+        Offset       : UInt := 0;
+        Level        : UInt := 0;
+        Continue     : Boolean := Width > 1 and then Height > 1;
+    begin
+        while Continue and then Level < Header.Mip_Map_Count loop
+            Mip_Size := ((Width + 3) / 4) * ((Height + 3) / 4) * Block_Size;
+            --  Load Compressed_Tex_Image_2D into the 2D texture
+            Targets.Texture_2D.Load_Compressed (Int (Level), Format,
+                              Int (Width), Int (Height), Int (Mip_Size),
+                              Image_Source (DDS_Data (Offset)'Address));
+
+            Continue :=  Width > 1 and then Height > 1;
+            if Continue then
+                Level := Level + 1;
+                Offset := Offset + Mip_Size;
+                Width := Width / 2;
+                Height := Height / 2;
+                --  Deal with Non-Power-Of-Two textures.
+                if Width < 1 then
+                    Width := 1;
+                end if;
+                if Height < 1 then
+                    Height := 1;
+                end if;
+            end if;
+        end loop;
+    end Load_Mipmaps;
+
+    --  ------------------------------------------------------------------------
+
+    function Test_DW_Four_CC (DW_Four_CC : String) return boolean is
+    begin
+        return DW_Four_CC = "DXT1" or else
+          DW_Four_CC =  "DXT3" or else
+          DW_Four_CC =  "DXT5";
+    end Test_DW_Four_CC;
+
+    --  ------------------------------------------------------------------------
+
+  use GL.Types;
+        use IO;
+
+        File_ID         : Byte_IO.File_Type;
+        Buffer_Length   : GL.Types.UInt := 0;
+        Bytes_Read      : GL.Types.UInt := 0;
+        Dump            : GL.Types.UByte;
+        Texture_ID      : GL.Types.UInt := 0;
+        Header          : tDDS_Header;
+begin
+    Byte_IO.Open (File_ID, Byte_IO.In_File, File_Name);
+    --  Determine file length
+    while not Byte_IO.End_Of_File (File_ID) loop
+        Buffer_Length := Buffer_Length + 1;
+        Read_UByte (File_ID, Dump);
+    end loop;
+
+    Byte_IO.Reset (File_ID);
+    declare
+        Input_Buffer : tUByte_Array (1 .. Buffer_Length);
```

Comment:
Reading the content into a buffer and then copying it from there with the `Load_DSS_*` functions is horribly inefficient. Use `Ada.Stream_IO` together with the `'Read` attribute instead to directly read the header and the content into appropriate variables.